src/System.Text.Json/src/System/Text/Json/JsonReaderState.cs

```diff
@@ -0,0 +1,110 @@
+        // reader does a linear scan, we only need to keep a single path as we go through the data.
+        // This is primarily used as an optimization to avoid having to allocate a stack object for
+        // depths up to 64 (which is the default max depth).
+        internal ulong _stackFreeContainer;
+        internal long _lineNumber;
+        internal long _lineBytePosition;
+        internal long _bytesConsumed;
+        internal int _currentDepth;
+        internal int _maxDepth;
+        internal bool _inObject;
+        internal bool _isNotPrimitive;
+        internal JsonTokenType _tokenType;
+        internal JsonReaderOptions _readerOptions;
+        internal Stack<JsonTokenType> _stack;
+        internal SequencePosition _sequencePosition;
+
+        /// <summary>
+        /// Returns the current <see cref="SequencePosition"/> within the provided UTF-8 encoded
+        /// input ReadOnlySequence&lt;byte&gt;. If the <see cref="JsonUtf8Reader"/> was constructed
+        /// with a ReadOnlySpan&lt;byte&gt; instead, this will always return a default <see cref="SequencePosition"/>.
+        /// </summary>
+        public SequencePosition Position => _sequencePosition;
+
+        /// <summary>
+        /// Returns the total amount of bytes consumed by the <see cref="JsonUtf8Reader"/> so far
+        /// for the given UTF-8 encoded input text.
+        /// </summary>
+        public long BytesConsumed => _bytesConsumed;
+
+        /// <summary>
+        /// Constructs a new <see cref="JsonReaderState"/> instance.
+        /// </summary>
+        /// <param name="maxDepth">Sets the maximum depth allowed when reading JSON, with the default set as 64.
+        /// Reading past this depth will throw a <exception cref="JsonReaderException"/>.</param>
+        /// <param name="commentHandling">Defines how the <see cref="JsonUtf8Reader"/> should handle comments when reading through the JSON.
+        /// By default, the <see cref="JsonUtf8Reader"/> treats comments within the JSON as invalid.</param>
+        /// <exception cref="ArgumentException">
+        /// Thrown when the max depth is set to a non-positive value (&lt;= 0)
+        /// </exception>
+        /// <remarks>
+        /// An instance of this state must be passed to the <see cref="JsonUtf8Reader"/> ctor with the JSON data.
+        /// Unlike the <see cref="JsonUtf8Reader"/>, which is a ref struct, the state can survive
+        /// across async/await boundaries and hence this type is required to provide support for reading
+        /// in more data asynchronously before continuing with a new instance of the <see cref="JsonUtf8Reader"/>.
+        /// </remarks>
+        public JsonReaderState(int maxDepth = StackFreeMaxDepth, JsonCommentHandling commentHandling = JsonCommentHandling.Default)
+        {
+            if (maxDepth <= 0)
+                ThrowHelper.ThrowArgumentException_MaxDepthMustBePositive();
```

Comment:
Doesn't this result in ThrowHelper always being the thing that reports as the top of the stack?  If you want to use a helper type for consistent formatting, that's fine, but I'd rather see it return the exception to be thrown from here, so the `throw` keyword is visible, indiciating control flow.

Reply:
Methods that conditionally throw exceptions on some paths can still be inlined, as can methods with complex flow where every path ultimately throws.

Simple methods that unconditionally throw won't be inlined w/o AggressiveInlining.

The benefits of the throw helper pattern are that it can save code size if the same exception can be thrown from multiple places within a method or across multiple methods, and that it also may reduce some of the time and space overhead the throwing path may impose on non-throwing paths in a method (say the method prolog).

Using a throw helper for just one site in one method may not provide any benefit -- you'd have to look at the non-exceptional codegen paths with and without a helper to see.

Reply:
Here are the permutations I tried.

**Using `throw new` versus using `throw helper`:**
1) Neither reader nor the state ctor marked with aggressively inlined
2) State ctor inlined
3) Reader ctor inlined
4) Both reader and state ctor marked with aggressively inlined

There was **no performance difference** (in this case) for any permutation.

However, I noticed that the StateCtor produced smaller codegen for the throw helper case (but after the return instruction, so likely doesn't matter):
https://www.diffchecker.com/diff
![image](https://user-images.githubusercontent.com/6527137/47971387-c6f35080-e045-11e8-8b87-edb2443e808b.png)

**Conclusion:**
> Using a throw helper for just one site in one method may not provide any benefit

Yep, in this case, the throw helper does not help, especially since this exception is only thrown in one place within this method and across all methods. I will explicitly use `throw` here.

1) No performance difference
ReaderCtor: 45.62 ns / 44.99 ns
StateCtor: 21.31 ns / 22.28 ns

2) No performance difference
ReaderCtor: 44.16 ns / 44.87 ns
StateCtor: 19.69 ns / 20. 11 ns

3) No performance difference
ReaderCtor: 27.50 ns / 27.82 ns
StateCtor: 21.84 ns / 21.73 ns

4) No performance difference
ReaderCtor: 28.36 ns / 27.40 ns
StateCtor: 20.96 ns / 20.01 ns